map/bookmark_manager.cpp

```diff
@@ -96,6 +103,115 @@ BookmarkManager::SharingResult GetFileForSharing(BookmarkManager::KMLDataCollect
+    }
+  }
+  catch (Reader::Exception const & ex)
+  {
+    LOG(LERROR, ("Error reading file:", filePath, ex.Msg()));
+    return false;
+  }
+  return true;
+}
+
+
+std::string buildIndexFile(std::vector<std::string> filesForIndex)
+{
+  std::string filePath = base::JoinPath(GetPlatform().TmpDir(), "doc.kml");
+  FileWriter fileWriter(filePath);
+  std::string content = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n"
+      "<kml xmlns=\"http://earth.google.com/kml/2.0\">\n"
+      "\n"
+      "<Document>\n"
+      "<name>Organic Maps backup</name>";
+  for (auto const & fileName : filesForIndex)
+  {
+    content.append("<NetworkLink><Link><href>");
+    content.append(fileName);
+    content.append("</href></Link>");
+    content.append("</NetworkLink>");
+  }
+  content.append("</Document>\n"
+             "</kml>");
+  fileWriter.Write(content.c_str(), content.length());
+  return filePath;
+}
+
+BookmarkManager::SharingResult exportMultipleFiles(BookmarkManager::KMLDataCollectionPtr collection)
+{
+  auto const kmzFileName = "OrganicMapsBackup";
+  auto kmzFilePath = base::JoinPath(GetPlatform().TmpDir(), kmzFileName + std::string{kKmzExtension});
+  zip::Create(kmzFilePath);
+  auto const filesDir = "files/";
+  base::DeleteFileX(kmzFilePath);
+  zip::File zip = zipOpen(kmzFilePath.c_str(), APPEND_STATUS_CREATE);
+  SCOPE_GUARD(kmzFileCloseGuard, std::bind(&zipClose, zip, nullptr));
+  int suffix = 0;
+  kml::GroupIdCollection categoriesIds;
+  for (auto const & kmlToExport : *collection)
+    categoriesIds.push_back(kmlToExport.second->m_categoryData.m_id);
+  std::vector<std::string> filesForIndex;
+  if (!zip) {
+    return {std::move(categoriesIds), BookmarkManager::SharingResult::Code::ArchiveError, "Could not create archive."};
+  }
+  for (auto const & kmlToExport : *collection)
+  {
+    std::string fileName = getFileNameForExport(kmlToExport);
+
+    auto const kmlPath = base::JoinPath(GetPlatform().TmpDir(), fileName + std::string{kKmlExtension});
+    SCOPE_GUARD(kmlFileGuard, std::bind(&base::DeleteFileX, kmlPath));
+
+    std::string fileNameInArchive = fileName;
+    if (!strings::IsASCIIString(fileNameInArchive))
+    {
+      fileNameInArchive = "OrganicMaps_" + std::to_string(suffix) + ".kml";
+      suffix++;
+    }
+    auto const filePathInArchive = filesDir + fileNameInArchive + std::string{kKmlExtension};
+    if (!SaveKmlFileSafe(*kmlToExport.second, kmlPath, KmlFileType::Text))
+      continue;
+    if (!addFileToArchive(kmlPath, filePathInArchive, zip))
+      continue;
+    filesForIndex.push_back(filePathInArchive);
+  }
+  auto indexFilePath = buildIndexFile(filesForIndex);
+  SCOPE_GUARD(indexFileGuard, std::bind(&base::DeleteFileX, indexFilePath));
+  addFileToArchive(indexFilePath, "doc.kml", zip);
+  return {std::move(categoriesIds), std::move(kmzFilePath)};
+}
+
+
+BookmarkManager::SharingResult GetFileForSharing(BookmarkManager::KMLDataCollectionPtr collection)
+{
+  if (collection->size() == 1)
```

Comment:
Can this `if` be avoided and one common implementation be reused that is based on `CreateZipFromFiles` method?

Reply:
It is possible, but for this case we need to have `if` on level when we decide do we need to create index file or not (or even for single file generate kmz with single kml in 'files' folder + index) + error handling is different (error if we can't save single file, skip for this error on multiple files) - so not ideal too. I reworked methods to use the same zip creation, what do you think?

Reply:
As discussed below, separate function is file.